### PR TITLE
ArgumentSyntax parsing modifications

### PR DESCRIPTION
**Pull request by alexanderkozlenko**
1. ArgumentSyntax doesn't affect application lifecycle via Environment.Exit / Environment.FailFast methods when parses arguments
2. ArgumentSyntax provides abilities to check if help is requested or there are some errors in arguments, this is important after modification no. 1
3. ArgumentSyntax provides full list of errors via AggregateException when errors are not handled by parser

### CODE REVIEW

**Comment by krwq** [comment]
LGTM

**Comment by alexanderkozlenko** [comment]
Maybe there should be two ways to parse arguments: Parse with throwing exception and TryParse with ability to get list of discovered errors?

**Comment by KrzysztofCwalina** [comment]
@terrajobst, could you review this PR?

**Comment by alexanderkozlenko** [comment]
I apologize, it turns out I broke additional validation support with commit 82869d6fb05a2aa64c86c4803cbfcf3e102c1c8d, fixed in commit b3d8735be5cabf4a0ddfbea32d3af04ca08a95d2

**Comment by terrajobst** [comment]
How is your change different from setting `HandleErrors` to `false`.

**Comment by alexanderkozlenko** [comment]
There are two scenarios to deal with errors:

1. ```HandleErrors = false```. Parser doesn't print discovered errors to console and throws exception if any.
2. ```HandleErrors = true```.  Parser prints discovered errors to console.

The major purpose of these modifications is to prevent termination of an application by parser in scenario 2. From my side this behavior is a bit confusing and unexpected (even if call to ```Environment.FailFast``` will be replaced with ```Environment.Exit```). Scenario 1 allows you to catch exception, do something, and then exit application. Scenario 2 doesn't allow this.

The minor purpose is to reveal all discovered errors in scenario 1 via ```AggregateException```.

**Comment by KrzysztofCwalina** [comment]
@terrajobst, are you fine with @alexanderkozlenko's response?

File: src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs [hunk 1]
```diff
@@ -52,12 +53,7 @@ private void Validate()
                 var helpText = GetHelpText();
                 Console.Error.Write(helpText);
 
-                // TODO: This should use Environment.Exit(0) but this API isn't available yet.
```
Comment (livarcocc) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: With this change, how will the caller of Parse know that the arguments passed where asking for help and that the execution of the program should stop?

File: src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs [hunk 2]
```diff
@@ -94,21 +93,12 @@ private bool IsHelpRequested()
                                    string.Equals(a, @"--help", StringComparison.Ordinal));
         }
 
-        public void ReportError(string message)
+		public void ReportError(string message)
```
Comment (khellang) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: Tabs :scream:

File: src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs (hunk 1 above)
Comment (alexanderkozlenko) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: Caller can use method ```public bool IsHelpRequested()```

Comment (livarcocc) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: But you are changing the external behavior of the method. There are people using this already who are expecting this to stop executing when help is requested. If you want to do this, you probably need to introduce an overload or a parameter to not go through this code. And then, new callers can use that. This way you won't break existing users of the API.

Comment (alexanderkozlenko) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: You're completely right, this modification is a breaking change for external behavior of Parse method. However the library is experimental and hasn't been shipped with .NET framework or via NuGet. Thus the main idea was to change this quite unexpected external behavior of the method during experimental stage.

By the way, all mentioned in [README](https://github.com/dotnet/corefxlab/tree/master/src/System.CommandLine#why-a-new-library) command line parser libraries don't attempt to terminate application during parsing for any reason.

File: src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs (hunk 2 above)
Comment (alexanderkozlenko) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: Fixed in commit 5b0a0178151f3946fa552c002056c43fb25a20ef

File: src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs (hunk 1 above)
Comment (livarcocc) [inline src/System.CommandLine/src/System/CommandLine/ArgumentSyntax.cs]: All right. Makes sense. We can make sure we are depending on a particular version of commandline and only depend on the new once once we change our code to handle the help requested properly.

**Comment by alexanderkozlenko** [comment]
It seems that PR #471 and referenced issue are about the same problem.

**Comment by daveaglick** [comment]
+1 for merging this PR (also, I prefer this approach over the one in #471).

Within 10 minutes of working with the library (and after figuring out the dependencies as resolved in #489) I had a "WTF?" moment when my app started crashing after requesting the help message. Even though this is a breaking change I'm with @alexanderkozlenko and think it should be the responsibility of the caller to decide what, if anything, should be done after a parsing failure or help request.

**Comment by alexanderkozlenko** [comment]
Is there any resolution, suggestions, or comments for this PR? :)

**Comment by daveaglick** [comment]
@alexanderkozlenko Any clarification on why this was closed? I thought it was a solid change.

The current `ReportError()` method still calls `Environment.Exit(1)`, which isn't great, IMO. Means we still can't handle CLI errors our own way (send to logging framework, output extra information, etc.)

**Pull request closed without merging**